src/lib/spinel/radio_spinel.cpp

```diff
@@ -997,10 +997,15 @@ otError RadioSpinel::AddSrcMatchShortEntry(uint16_t aShortAddress)
 {
     otError error;
 
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0
+    EXPECT_NO_ERROR(error = (mSrcMatchShortEntryCount < OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES)
+                                ? OT_ERROR_NONE
+                                : OT_ERROR_NO_BUFS);
+#endif
+
     EXPECT_NO_ERROR(error = Insert(SPINEL_PROP_MAC_SRC_MATCH_SHORT_ADDRESSES, SPINEL_DATATYPE_UINT16_S, aShortAddress));
 
 #if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0
-    assert(mSrcMatchShortEntryCount < OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES);
```

Comment:
The value of [OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES is set to OPENTHREAD_CONFIG_MLE_MAX_CHILDREN](https://github.com/openthread/openthread/blob/main/src/lib/spinel/openthread-spinel-config.h#L98) by default. Theoretically, the number of the source match addresses will not exceed OPENTHREAD_CONFIG_MLE_MAX_CHILDREN. You may need to check why the number of the source match addresses exceeds OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES.

Reply:
Hi, @zhanglongxia @abtink  Actually, we did not run Thread as the upper layer over spinel, we support the Zigbee stack over spinel host. For thread, yes, there is a mechanism to confirm that only the size of `MLE_MAX_CHILDREN` will be used as the entry count of the matching table. But, if we run other stack over spinel, should we also assert here? Or just return an error code to let the upper layer to handle it?

Reply:
I think if you are running other stacks over the spinel host, you should define new vendor Spinel properties to set the source match address, rather than using Thread's source match address setting methods.

Reply:
I think the API `RadioSpinel::AddSrcMatchShortEntry` is intended to add addresses to the matching table to determine if the pending bit should be set in the ACK frame. It appears to be an API defined in the Spinel layer. In the Zigbee stack, we also need to add addresses to the matching table. Defining additional functions for this would likely result in similar functionality.

So, the main question is: why can’t we handle this error more gracefully (e.g., by returning an error instead of asserting)?

Reply:
> why can’t we handle this error more softly (e.g., by returning an error instead of asserting)?

@zwx1995esp, this is not about `AddSrcMatchShortEntry()` failing and returning an error (which it can, and this is perfectly okay and handled properly).

This `assert` is part of the code related to the `RCP_RESTORATION` feature, which aims to restore the RCP (from the spinel host) if it gets reset. This avoids resetting the entire OT stack on the host due to an RCP reset. Here, the code is trying to allocate an entry to save the already added "short MAC" entry on RCP so it can restore the RCP (populate the src match table again) if the RCP is reset and need restoration. I mentioned above that this change can break the RCP restoration behavior.

---


Here are my suggestions:

-   If you want to use radio spinel for another protocol, try to determine a reasonable upper estimate for the number of nodes and source address match entries that may be used. Update `OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES` in `radio_spinel` accordingly to accommodate not only what the Thread stack would use but also the other stack. (We already did this for Thread, which is why `OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES` is set to `MAX_CHILDREN` by default.)
-   Alternatively, you can disable the `RCP_RESTORATION` feature (set `OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT` to zero). Not using the `RCP_RESTORATION` feature may be a better approach, as I would expect other issues could also cause RCP restoration to malfunction when the radio/RCP is shared between multiple stacks.

Reply:
> This `assert` is part of the code related to the `RCP_RESTORATION` feature, which aims to restore the RCP (from the spinel host) if it gets reset. This avoids resetting the entire OT stack on the host due to an RCP reset. Here, the code is trying to allocate an entry to save the already added "short MAC" entry on RCP so it can restore the RCP (populate the src match table again) if the RCP is reset and need restoration. I mentioned above that this change can break the RCP restoration behavior.
> 
Hi @abtink  AFAK: the line:
```
 EXPECT_NO_ERROR(error = Insert(SPINEL_PROP_MAC_SRC_MATCH_SHORT_ADDRESSES, SPINEL_DATATYPE_UINT16_S, aShortAddress));
```
will add the address to RCP, we move the checking logic in front of  this line. Before inserting the short address onto RCP, we first check if there is an empty entry in the array to store the address. If not, only raise an error and do not add it onto RCP.

Should this logic be reasonable?

Reply:
Also it seems the macro check surrounding `#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0` should be useless. We can also remove it.

Reply:
Okay, I see how this can work, but the downside is that it adds an additional restriction to the RCP host (`radio_spinel`). It fails to add to the "source match table" if there is no space to save the address for RCP restoration, rather than letting the RCP dictate this. This indirectly penalizes Thread stack behavior when the RCP is shared with another stack (which feels undesirable).

That said, I am okay with including this change but highly recommend that if you intend to share the radio, to ensure to set `OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES`  to a large enough number to account for the number of "source match table" entries used by the other stack, along with the maximum needed for Thread operation.

> Also, it seems the surrounding macro checking `#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0` should be useless. We can also remove it.

I don't think we can. The `mSrcMatchShortEntryCount` and `mSrcMatchExtEntryCount` variables are only defined and available under `#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0` and are not needed or used otherwise (when RCP restoration is not enabled).

Reply:
Ok, Thanks abtink. I also agree with your suggestion, to set OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES to a large enough number should solve this problem fundamentally. We will set a larger number of entries for zigbee stack.